web3/_utils/module.py

```diff
@@ -34,18 +34,19 @@ def attach_modules(
                 "already has an attribute with that name"
             )
 
-        if issubclass(module_class, Module):
-            # If the `module_class` inherits from the `web3.module.Module` class, it has access to
-            # caller functions internal to the web3.py library and sets up a proper codec. This
-            # is likely important for all modules internal to the library.
-            if w3 is None:
-                setattr(parent_module, module_name, module_class(parent_module))
-                w3 = parent_module
-            else:
-                setattr(parent_module, module_name, module_class(w3))
-        else:
-            # An external `module_class` need not inherit from the `web3.module.Module` class.
-            setattr(parent_module, module_name, module_class)
+        # The parent module is the ``Web3`` instance on first run of the loop
+        if type(parent_module).__name__ == 'Web3':
+            w3 = parent_module
+
+        try:
+            # Modules that need access to the ``Web3`` instance may accept the instance as the first
+            # arg in their ``__init__()`` method. This is the case for any module that inherits from
+            # ``web3.module.Module``.
+            # e.g. def __init__(self, w3):
+            setattr(parent_module, module_name, module_class(w3))
+        except TypeError:
```

Comment:
Just from my own ignorance, when does this `TypeError` get raised?

Reply:
This was a bit tough to navigate but this is the current flow:

1. Assume the first argument to a module's `__init__()`, if there is one, is a web3 instance
2. If it is, we pass it in and the module may make use of the instance in case it needs to make internal calls
3. If there is no argument to be accepted in the `__init__()`, it throws a type error and we catch it here and, instead, initialize the module without any arguments. This is a module that makes no use of the web3 instance.

I had some trouble with trying to read the first argument on the class because the class isn't initialized yet at this point. We are passing in a static class and these lines are actually where they get initialized. I think a separate (and maybe better approach?) is to create a utility method that reads the static class, more specifically the `__init__()` of the static class, and sees if there is an argument passed in. If there is, we can initialize it with `module_class(w3)`, if there is not then we initialize it as `module_class()`.

100% open to changing this if you see a better way. Thoughts?